lib/Net/Async/Blockchain/ETH.pm

```diff
@@ -415,73 +434,81 @@ async sub _set_transaction_type {
 
 =head2 _check_contract_transaction
 
-We need to identify what are the transactions that have a contract as
-destination, once we found we change:
-
-currency => the contract symbol
-amount => tokens
-to => address that will receive the tokens
-contract => the contract address
+For now this method just check the ERC20 contracts
 
 =over 4
 
-=item * L<Net::Async::Blockchain::Transaction>
+=item * C<$client> - a L<Net::Async::Blockchain::ETH> instance
 
 =back
 
-hashref from an array of L<Net::Async::Blockchain::Transaction>
+Returns an array of hashrefs with fields matching attributes in L<Net::Async::Blockchain::Transaction>
 
 =cut
 
 async sub _check_contract_transaction {
-    my ($self, $transaction) = @_;
-
-    # ERC20 transfer
-    # transfer(address _to, uint256 _value)
-    # signature = 4 bytes
-    # address = 32 bytes
-    # amount = 32 bytes
-    # total = 68 bytes, characters = 136
-    # including "0x" = 138
-    if (length($transaction->data) >= 138 && substr($transaction->data, 0, 10) eq substr(TRANSFER_SIGNATURE, 0, 10)) {
-        my $address = substr($transaction->data, 10, 64);
-        my $amount  = substr($transaction->data, 74, 64);
-
-        return undef unless $address && $amount;
-
-        my $contract_address = $transaction->to;
-
-        my $hex_symbol = await $self->rpc_client->call({
-                data => SYMBOL_SIGNATURE,
-                to   => $contract_address
-            },
-            "latest"
-        );
+    my ($self, $transaction, $receipt) = @_;
+
+    my $logs = $receipt->{logs};
+
+    return () unless @$logs;
+
+    my @transactions;
+
+    return () unless $receipt->{status} && hex($receipt->{status}) == 1;
+
+    for my $log ($logs->@*) {
+        my @topics = $log->{topics}->@*;
+        if (@topics && $topics[0] eq TRANSFER_EVENT_SIGNATURE) {
+            my $transaction_cp = $transaction->clone();
+
+            my $address = $log->{address};
+            my $amount  = Math::BigFloat->from_hex($log->{data});
+
+            my $hex_symbol = await $self->rpc_client->call({
+                    data => SYMBOL_SIGNATURE,
+                    to   => $address
+                },
+                "latest"
+            );
+
+            my $symbol = $self->_to_string($hex_symbol);
+            next unless $symbol;
+
+            $transaction_cp->{currency} = $symbol;
+
+            my $decimals = await $self->rpc_client->call({
+                    data => DECIMALS_SIGNATURE,
+                    to   => $address
+                },
+                "latest"
+            );
+
+            if ($decimals) {
+                # default size 64 + `0x`
+                next unless length($decimals) == 66;
+                my $bg_decimals = Math::BigInt->from_hex($decimals);
+                # decimal places can't go over 18
```

Comment:
then we're expecting a value up to `0x000000000000000000000000000000000000000000000000000000000000000012`? can add better sanity checking here, for example by rejecting anything with less than 60 leading `0`